In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'/Users/sofiakazeko/Jupiter/Dec-2024 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_test = pd.DataFrame()

In [6]:
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
#Finding Missing Values
df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
#dataframe w/o missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
#finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups 

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
#Export df_prods_clean dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv')) 


# Task

In [26]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

After reviewing the values in each column, no anomalous or out-of-range values were identified. All values seem to be within a reasonable and interpretable range for their respective columns/variables.


In [28]:
#check for mixed types
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != type(df_ords[col].iloc[0])).any()
    if weird:
        print(col)



Unnamed: 0
order_id
user_id
order_number
orders_day_of_week
order_time
days_since_prior_order


In [29]:
# Check the data types of all columns
print(df_ords.dtypes)


Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_time                  int64
days_since_prior_order    float64
dtype: object


#no column with mixed data type detected

# 5- finding missing values

In [31]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

#The days_since_prior_order column has 206,209 missing values, which is about 6.02% of the total rows. This could be because some rows represent first-time orders with no prior data, or it might be due to the 30-day limit where orders placed more than 30 days apart are not recorded. Alternatively, the missing values could also be caused by errors or incomplete data.
First - we can check if each user id has the missing value in this column and how many.

In [41]:
# Count the number of missing values for each user_id
missing_per_user = df_ords[df_ords['days_since_prior_order'].isna()].groupby('user_id').size()


In [42]:
print(missing_per_user)


user_id
1         1
2         1
3         1
4         1
5         1
         ..
206205    1
206206    1
206207    1
206208    1
206209    1
Length: 206209, dtype: int64


In [43]:
# Check if all user_ids with missing values have exactly one missing value
if (missing_per_user == 1).all():
    print("Each user_id has only one missing value, likely representing the first order.")
else:
    print("Some user_ids have more than one missing value, further investigation needed.")

Each user_id has only one missing value, likely representing the first order.


6. There’s no need to remove or input themissing values, as each unique user_id has only one missing value in the days_since_prior_order column, which corresponds to their very first order. 

In [44]:
#finding duplicates
df_dupsords = df_ords[df_ords.duplicated()]

In [45]:
df_dupsords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

#7&8 There are no duplicate values found

#before exporting I would like to remove unnecessary column Unnamed since it's duplicating numbering. 

In [51]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_time  days_since_prior_order  
0           8                     NaN  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

In [57]:
#dropping Unnamed:0 column without overwrititng old dataframe.
df_ords_clean= df_ords.drop(columns=['Unnamed: 0'])


In [56]:
df_ords_clean.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [59]:
#Export df_ords_clean dataframe
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv')) 


In [60]:
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [61]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  order_time  \
0         2539329        1             1                   2           8   
1         2398795        1             2                   3           7   
2          473747        1             3                   3          12   
3         2254736        1             4                   4           7   
4          431534        1             5                   4          15   
...           ...      ...           ...                 ...         ...   
3421078   2266710   206209            10                   5          18   
3421079   1854736   206209            11                   4          10   
3421080    626363   206209            12                   1          12   
3421081   2977660   206209            13                   1          12   
3421082    272231   206209            14                   6          14   

         days_since_prior_order  new_customer  
0                           NaN          True  
1                          15.0         False  
2                          21.0         False  
3                          29.0         False  
4                          28.0         False  
...                         ...           ...  
3421078                    29.0         False  
3421079                    30.0         False  
3421080                    18.0         False  
3421081                     7.0         False  
3421082                    30.0         False  

[3421083 rows x 7 columns]

In [62]:
#Export df_ords_clean dataframe
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv')) 